This notebook is used for exploring the data to process.

In [1]:
import pandas as pd

In [7]:
data = pd.read_csv('../data/raw/historical_stock.csv', index_col=0, header=[0,1])

df = pd.DataFrame(data)

df.head(10)

Ticker        IBE.MC                                               ITX.MC  \
Price           Open      High       Low     Close      Volume       Open   
Date                                                                        
2023-09-18  9.753249  9.819598  9.691324  9.779789   6808689.0  32.694881   
2023-09-19  9.766519  9.885947  9.766519  9.850561   5080644.0  32.044418   
2023-09-20  9.877099  9.987680  9.859406  9.961141   7318422.0  31.999242   
2023-09-21  9.872676  9.912485  9.691323  9.739979   7508414.0  32.107662   
2023-09-22  9.678053  9.810751  9.616128  9.695746  18761000.0  31.926972   
2023-09-25  9.673631  9.682477  9.470162  9.527664   5548543.0  31.710148   
2023-09-26  9.465737  9.642667  9.430351  9.567472   8082530.0  31.511395   
2023-09-27  9.540934  9.558627  9.390543  9.403813  26413886.0  31.583669   
2023-09-28  9.425930  9.492278  9.319772  9.350735   9329256.0  31.475256   
2023-09-29  9.399390  9.496701  9.355157  9.372850  19508913.0  32.234136   

Ticker                                                  ...       SAN  \
Price            High        Low      Close     Volume  ...      Open   
Date                                                    ...             
2023-09-18  32.712950  32.107658  32.152828  1772778.0  ...  3.334987   
2023-09-19  32.378684  31.719186  31.809526  1266608.0  ...  3.325900   
2023-09-20  32.469023  31.881800  32.333511  1930625.0  ...  3.389511   
2023-09-21  32.369656  31.683052  32.053455  3591947.0  ...  3.316813   
2023-09-22  31.954073  31.719185  31.899870  1388392.0  ...  3.362249   
2023-09-25  32.089587  31.565600  31.791456  1641218.0  ...  3.298639   
2023-09-26  31.872762  31.430086  31.592703  2697237.0  ...  3.253203   
2023-09-27  31.800492  31.384918  31.619806  2850699.0  ...  3.316813   
2023-09-28  32.207030  31.357810  32.134754  2116787.0  ...  3.407685   
2023-09-29  32.432890  31.737254  31.881802  2495726.0  ...  3.471295   

Ticker                                                   AIR.PA              \
Price           High       Low     Close     Volume        Open        High   
Date                                                                          
2023-09-18  3.344075  3.289551  3.307726  2511400.0  124.020145  124.114535   
2023-09-19  3.353162  3.316813  3.344075  1933400.0  122.170228  123.000809   
2023-09-20  3.416772  3.362249  3.371336  2590300.0  122.207976  123.076304   
2023-09-21  3.371336  3.316813  3.334988  1984800.0  121.075374  121.339647   
2023-09-22  3.371336  3.307726  3.325900  2095300.0  117.677565  118.168357   
2023-09-25  3.307726  3.262290  3.298639  2514200.0  116.846995  117.677569   
2023-09-26  3.289552  3.244116  3.253203  3137100.0  115.865405  117.941843   
2023-09-27  3.316813  3.271377  3.298639  3930500.0  116.582714  117.186776   
2023-09-28  3.471295  3.407685  3.444033  9387000.0  117.224526  119.697381   
2023-09-29  3.480382  3.407685  3.416772  3795000.0  119.300969  121.075380   

Ticker                                         
Price              Low       Close     Volume  
Date                                           
2023-09-18  122.453387  122.736526   537506.0  
2023-09-19  121.811566  122.019211   553728.0  
2023-09-20  121.603921  122.358994   637125.0  
2023-09-21  118.470383  118.470383  1341077.0  
2023-09-22  116.677101  116.922493  1211999.0  
2023-09-25  115.299104  116.054169  1130558.0  
2023-09-26  114.260878  116.639351  1064435.0  
2023-09-27  115.978659  117.035759   828813.0  
2023-09-28  116.979133  119.621872  1028866.0  
2023-09-29  118.829051  119.848396  1177842.0  

[10 rows x 25 columns]

In [17]:
df.info() 
df.isnull().sum()

<class 'pandas.DataFrame'>
Index: 774 entries, 2023-09-18 to 2026-09-17
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   (IBE.MC, Open)    764 non-null    float64
 1   (IBE.MC, High)    764 non-null    float64
 2   (IBE.MC, Low)     764 non-null    float64
 3   (IBE.MC, Close)   764 non-null    float64
 4   (IBE.MC, Volume)  764 non-null    float64
 5   (ITX.MC, Open)    764 non-null    float64
 6   (ITX.MC, High)    764 non-null    float64
 7   (ITX.MC, Low)     764 non-null    float64
 8   (ITX.MC, Close)   764 non-null    float64
 9   (ITX.MC, Volume)  764 non-null    float64
 10  (SIE.DE, Open)    760 non-null    float64
 11  (SIE.DE, High)    760 non-null    float64
 12  (SIE.DE, Low)     760 non-null    float64
 13  (SIE.DE, Close)   760 non-null    float64
 14  (SIE.DE, Volume)  760 non-null    float64
 15  (SAN, Open)       752 non-null    float64
 16  (SAN, High)       752 non-null    float64
 1

Ticker  Price 
IBE.MC  Open      10
        High      10
        Low       10
        Close     10
        Volume    10
ITX.MC  Open      10
        High      10
        Low       10
        Close     10
        Volume    10
SIE.DE  Open      14
        High      14
        Low       14
        Close     14
        Volume    14
SAN     Open      22
        High      22
        Low       22
        Close     22
        Volume    22
AIR.PA  Open      10
        High      10
        Low       10
        Close     10
        Volume    10
dtype: int64

The *historical_stock.csv* file has 774 entries, from **2023-09-18** to **2026-09-17**. It is easyly observable that at least 10 days are null in every column, those days need to be assessed, as well as some other faulty registries clearly shown in the information above. 

In [25]:
null_dates = df.isnull()

null_dates = null_dates[null_dates.any(axis=1)]
null_dates = null_dates.index.tolist()
print(null_dates)
print(f"Number of null dates: {len(null_dates)}")
print(f"Percentage of null dates: {len(null_dates)/len(df)*100:.2f}%")

['2023-11-23', '2023-12-26', '2024-01-15', '2024-02-19', '2024-04-01', '2024-05-01', '2024-05-27', '2024-06-19', '2024-07-04', '2024-09-02', '2024-11-28', '2024-12-24', '2024-12-26', '2024-12-31', '2025-01-09', '2025-01-20', '2025-02-17', '2025-04-21', '2025-05-01', '2025-05-26', '2025-06-19', '2025-07-04', '2025-09-01', '2025-11-27', '2025-12-24', '2025-12-26', '2025-12-31', '2026-01-19', '2026-02-16', '2026-04-06', '2026-05-01', '2026-05-25', '2026-06-19', '2026-07-03', '2026-09-16', '2026-09-17']
Number of null dates: 36
Percentage of null dates: 4.65%


Being the percentage of null dates as low as $4.65\%$, these registries can be deleted without compromising the integrity of this study. 

In [26]:
df.dropna(inplace=True)